l1-contracts/contracts/bridge/L1NativeTokenVault.sol

```diff
@@ -0,0 +1,289 @@
+        }
+
+        (, bytes memory data1) = _token.staticcall(abi.encodeCall(IERC20Metadata.name, ()));
+        (, bytes memory data2) = _token.staticcall(abi.encodeCall(IERC20Metadata.symbol, ()));
+        (, bytes memory data3) = _token.staticcall(abi.encodeCall(IERC20Metadata.decimals, ()));
+        return abi.encode(data1, data2, data3);
+    }
+
+    ///  @inheritdoc IL1AssetHandler
+    function bridgeMint(
+        uint256 _chainId,
+        bytes32 _assetId,
+        bytes calldata _data
+    ) external payable override onlyBridge whenNotPaused returns (address _l1Receiver) {
+        // here we are minting the tokens after the bridgeBurn has happened on an L2, so we can assume the l1Token is not zero
+        address l1Token = tokenAddress[_assetId];
+        uint256 amount;
+        (amount, _l1Receiver) = abi.decode(_data, (uint256, address));
+        if (!L1_SHARED_BRIDGE.hyperbridgingEnabled(_chainId)) {
+            // Check that the chain has sufficient balance
+            require(chainBalance[_chainId][l1Token] >= amount, "NTV not enough funds 2"); // not enough funds
+            chainBalance[_chainId][l1Token] -= amount;
+        }
+        if (l1Token == ETH_TOKEN_ADDRESS) {
+            bool callSuccess;
+            // Low-level assembly call, to avoid any memory copying (save gas)
+            assembly {
+                callSuccess := call(gas(), _l1Receiver, amount, 0, 0, 0, 0)
+            }
+            require(callSuccess, "NTV: withdrawal failed, no funds or cannot transfer to receiver");
+        } else {
+            // Withdraw funds
+            IERC20(l1Token).safeTransfer(_l1Receiver, amount);
+        }
+        // solhint-disable-next-line func-named-parameters
+        emit BridgeMint(_chainId, _assetId, _l1Receiver, amount);
+    }
+
+    ///  @inheritdoc IL1AssetHandler
+    function bridgeRecoverFailedTransfer(
+        uint256 _chainId,
+        bytes32 _assetId,
+        address,
+        bytes calldata _data
+    ) external payable override onlyBridge whenNotPaused {
+        (uint256 _amount, address _depositSender) = abi.decode(_data, (uint256, address));
+        address l1Token = tokenAddress[_assetId];
+        require(_amount > 0, "y1");
+
+        if (l1Token == ETH_TOKEN_ADDRESS) {
+            bool callSuccess;
+            // Low-level assembly call, to avoid any memory copying (save gas)
+            assembly {
+                callSuccess := call(gas(), _depositSender, _amount, 0, 0, 0, 0)
+            }
+            require(callSuccess, "NTV: claimFailedDeposit failed, no funds or cannot transfer to receiver");
+        } else {
+            IERC20(l1Token).safeTransfer(_depositSender, _amount);
+            // Note we don't allow weth deposits anymore, but there might be legacy weth deposits.
+            // until we add Weth bridging capabilities, we don't wrap/unwrap weth to ether.
+        }
+
+        if (!L1_SHARED_BRIDGE.hyperbridgingEnabled(_chainId)) {
```

Comment:
NOT NIT: always follow [Checks Effects Interactions](https://fravoll.github.io/solidity-patterns/checks_effects_interactions.html)[Checks Effects Interactions](https://fravoll.github.io/solidity-patterns/checks_effects_interactions.html) pattern. 

In this function I am pretty sure it is possible to Drain the chain balance.